monitoring/transforming_exporter.go

```diff
@@ -0,0 +1,193 @@
+			return
+		}
+		op.dropLabels = append(op.dropLabels, toDrop...)
+	}
+}
+
+func makeMetricOperation() metricOperation {
+	return metricOperation{
+		labelNameMap:  make(map[string]string),
+		labelValueMap: make(map[string]LabelValueMapperFn),
+		constLabels:   make(map[string]string),
+		dropLabels:    make([]string, 0),
+	}
+}
+
+type transformingExporter struct {
+	baseExporter     metricexport.Exporter
+	metricOperations operations
+	dropUnmodified   bool
+}
+
+func (te *transformingExporter) ExportMetrics(ctx context.Context, data []*metricdata.Metric) error {
+	newData := make([]*metricdata.Metric, 0, len(data))
+	for _, datum := range data {
+		op, found := te.metricOperations[datum.Descriptor.Name]
+		if !found {
+			if !te.dropUnmodified {
+				newData = append(newData, datum)
+			}
+			continue
+		}
+		newMetric := &metricdata.Metric{Resource: datum.Resource}
+		newMetric.Descriptor = metricdata.Descriptor{
+			Name:        datum.Descriptor.Name,
+			Description: datum.Descriptor.Description,
+			Unit:        datum.Descriptor.Unit,
+			Type:        datum.Descriptor.Type,
+			LabelKeys:   datum.Descriptor.LabelKeys,
+		}
+		newMetric.TimeSeries = datum.TimeSeries
+		if op.newName != "" {
+			newMetric.Descriptor.Name = op.newName
+		}
+		if len(op.labelNameMap) > 0 || len(op.constLabels) > 0 {
+			newKeys := make([]metricdata.LabelKey, 0, len(datum.Descriptor.LabelKeys))
+			for _, key := range datum.Descriptor.LabelKeys {
+				newKey, found := op.labelNameMap[key.Key]
+				if found {
+					newKeys = append(newKeys, metricdata.LabelKey{Key: newKey, Description: key.Description})
+				} else {
+					newKeys = append(newKeys, key)
+				}
+			}
+			for k, _ := range op.constLabels {
+				newKeys = append(newKeys, metricdata.LabelKey{Key: k})
+			}
+			newMetric.Descriptor.LabelKeys = newKeys
+		}
+		if len(op.labelValueMap) > 0 || len(op.constLabels) > 0 {
+			newTimeSeries := make([]*metricdata.TimeSeries, 0, len(datum.TimeSeries))
+			for _, ts := range datum.TimeSeries {
+				newTS := &metricdata.TimeSeries{StartTime: ts.StartTime, Points: ts.Points, LabelValues: make([]metricdata.LabelValue, 0, len(ts.LabelValues))}
+				for i, lv := range ts.LabelValues {
+					fn, found := op.labelValueMap[datum.Descriptor.LabelKeys[i].Key]
+					if found {
+						newTS.LabelValues = append(newTS.LabelValues, metricdata.LabelValue{Value: fn(lv.Value), Present: lv.Present})
+					} else {
+						newTS.LabelValues = append(newTS.LabelValues, lv)
+					}
+				}
+				for _, v := range op.constLabels {
+					newTS.LabelValues = append(newTS.LabelValues, metricdata.LabelValue{Value: v, Present: true})
+				}
+				newTimeSeries = append(newTimeSeries, newTS)
+			}
+			newMetric.TimeSeries = newTimeSeries
+		}
+
+		newData = append(newData, newMetric)
+	}
+
+	return te.baseExporter.ExportMetrics(ctx, newData)
+}
+
+func NewTransformingExporter(baseExporter metricexport.Exporter, dropUnmodified bool, ops ...MetricOperation) metricexport.Exporter {
```

Comment:
I think this would work where your external wrapper would wrap both exporters and your SD exporter would be wrapped by this exporter. So, it functions as middleware, sorta like:

NewMultiExporter(promExporter, NewTransformingExporter(sdExporter))

WDYT about adding a NewWrappingExporter to support that use case explicitly (vs requiring custom code)?

Eventually, I think we'll want to replace all of this with OTel client libraries -- and their eco-system of processors, etc.

Reply:
> WDYT about adding a NewWrappingExporter to support that use case explicitly (vs requiring custom code)?

Yeah that will work, but the custom code, at least for now, is going to be very small and no way could conflict or out of sync with oss. The work of refactoring might not justify the value, if it is just to get derived gauge to work.

> Eventually, I think we'll want to replace all of this with OTel client libraries -- and their eco-system of processors, etc.

Does OT have some first class support for this kind of use case.